packages/react-native-sdk/src/hooks/useCallControls.tsx

```diff
@@ -57,17 +64,65 @@ export const useCallControls = () => {
 
   const publishVideoStream = useCallback(async () => {
     try {
-      if (currentVideoDevice) {
+      if (isOnline && currentVideoDevice) {
         const videoStream = await getVideoStream(currentVideoDevice.deviceId);
-        if (call) await call.publishVideoStream(videoStream);
+        if (call) {
+          await call.publishVideoStream(videoStream);
+        }
       }
     } catch (e) {
       console.log('Failed to publish video stream', e);
     }
-  }, [call, currentVideoDevice]);
+  }, [call, currentVideoDevice, isOnline]);
 
-  /* Attempt to republish video stream when app comes back to foreground */
-  useAppStateListener(publishVideoStream);
+  /** Refs to keep track of whether the user has published the track before going offline  */
+  const isAudioPublishedRef = useRef(false);
+  const isVideoPublishedRef = useRef(false);
+  isAudioPublishedRef.current = !isAudioMuted;
+  isVideoPublishedRef.current = !isVideoMuted;
+
+  /** Refs to be used for useEffect that does the rejoining flow when coming back offline */
+  const publishAudioStreamRef = useRef(publishAudioStream);
+  const publishVideoStreamRef = useRef(publishVideoStream);
+  publishAudioStreamRef.current = publishAudioStream;
+  publishVideoStreamRef.current = publishVideoStream;
+
+  /** Attempt to republish video stream when app comes back to foreground */
+  useAppStateListener(
+    isVideoPublishedRef.current ? publishVideoStream : undefined,
+  );
+
+  /**
+   * Effect to re-join to an existing call happens in case the user comes back online
+   */
+  useEffect(() => {
+    if (!call) return;
+    if (isOnline) {
```

Comment:
This file's cognitive complexity is high, making it difficult to read and comprehend. 
- Please go for early exits (`if(!isOnline)` over big if statements (`if(isOnline) {...}`).
- Merge if's where possible.
- We agreed on using async await in try-catch blocks over then,catch.
-